### PR TITLE
Added yamllint.

### LINKED ISSUE #693: [Feature Request] Add yamllint to lint the config files
# 🚀 Feature Request

Add [yamllint](https://yamllint.readthedocs.io/en/stable/index.html) to test suite to ensure config files are linted.

## Motivation

NA

## Pitch

NA

**Are you willing to open a pull request?** (See [CONTRIBUTING](../CONTRIBUTING.md))

yes

## Additional context

NA

### PR DESCRIPTION
**Pull request by shagunsodhani**
## Motivation

Add yamllint and fix linting for yaml files.
Closes #693 

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebookresearch/hydra/blob/master/CONTRIBUTING.md)?

Yes

## Test Plan

NA

## Related Issues and PRs

https://github.com/facebookresearch/hydra/issues/693

### CODE REVIEW

**Comment by shagunsodhani** [comment]
On running locally, the most common errors were:

* missing document start "---"  (document-start)
* no new line character at the end of file  (new-line-at-end-of-file)
* line too long (122 > 80 characters)  (line-length) (This could force some keys to be renamed so I am in favor of disabling this error).
* too many blank lines (1 > 0)  (empty-lines)
* wrong indentation: expected 2 but found 4  (indentation)
* trailing spaces  (trailing-spaces)

Let me know if there is any error we should disable. I will fix the remaining errors in the next commit.

Regarding testing, I am running yamllint just like flake8 - on the entire directory at once. 

I would have to ignore certain folders (since hydra jobs log their config as yaml files). I am thinking of ignoring any files in the `.hydra` directory.

**Comment by shagunsodhani** [comment]
Errors from CI: https://gist.github.com/shagunsodhani/c977bc5813791a533f715feb8c9c89af

Error Type/Frequency:

'missing document start "---"  (document-start)': 199,
'no new line character at the end of file  (new-line-at-end-of-file)': 141
'trailing spaces  (trailing-spaces)': 5
'too many spaces before colon  (colons)': 4
'truthy value should be one of [false, true]  (truthy)': 6
'too many blank lines (1 > 0)  (empty-lines)': 3
'line too long (125 > 80 characters)  (line-length)': 3
'line too long (94 > 80 characters)  (line-length)': 2
'line too long (111 > 80 characters)  (line-length)': 2
'line too long (122 > 80 characters)  (line-length)': 2
'line too long (101 > 80 characters)  (line-length)': 1
'line too long (82 > 80 characters)  (line-length)': 1
'line too long (89 > 80 characters)  (line-length)': 1
'line too long (90 > 80 characters)  (line-length)': 1
'line too long (97 > 80 characters)  (line-length)': 1
'wrong indentation: expected 2 but found 0  (indentation)': 1
'wrong indentation: expected 2 but found 4  (indentation)': 1
'wrong indentation: expected 4 but found 6  (indentation)': 1

**Comment by omry** [comment]
Thanks!
but your PR is now failing the lint checks :)

**Comment by shagunsodhani** [comment]
That is right. 

I am thinking of ignoring the `line too long (122 > 80 characters) (line-length)` as this could force some keys to be renamed so I am in favor of disabling this error. Others can be fixed. Does that sound right ?

**Comment by omry** [comment]
Certainly, we need to do what makes sense.
configure it to not bitch about line length.

**Comment by omry** [comment]
is there a corresponding yaml formatter?

**Comment by shagunsodhani** [comment]
No there isn't one. I will write a script for the fixing them for now.

Thanks

Shagun
https://shagunsodhani.com

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,3 +1,4 @@
+---
```
Comment (omry) [inline .circleci/config.yml:1]: ahh, can we disable this one?
I hate it, a lot.

File: .yamllint [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  braces: enable
+  brackets: enable
+  colons: enable
+  commas: enable
+  comments:
+    level: warning
+  comments-indentation:
+    level: warning
+  document-end: disable
+  document-start: enable
```
Comment (omry) [inline .yamllint:18]: here, disable here. I don't want to contaminate all the yaml files with this.

**Comment by omry** [comment]
Thanks!
Can you update the contributing guide?

File: noxfile.py [hunk 3]
```diff
@@ -216,6 +216,7 @@ def lint(session):
     )
     session.run("mypy", ".", "--strict", silent=SILENT)
     session.run("flake8", "--config", ".flake8")
+    session.run("yamllint", ".")
```
Comment (omry) [inline noxfile.py:219]: can you remove the installation of black from line 200?
I don't think it's needed.

Comment (shagunsodhani) [reply noxfile.py:219]: Will push once these changes pass on CI :)

**Comment by shagunsodhani** [comment]
Do we need a towncrier entry as well ?

File: website/docs/development/contributing.md [hunk 4]
```diff
@@ -39,6 +39,9 @@ The code need to pass verification by the following tools:
  - `flake8` : PEP8 compliance checker for Python, this includes copyright header verification
  - `isort .` : Ensure imports are sorted properly
  - `mypy --strict .` : Ensures code passes strict type checking
+ - `yamllint .` : Ensures that yaml files are syntactically correct and properly indented.
+
+`yamllint` will forbid the use of document start marker (`---`). When running locally, `yamllint` will run on all the yaml files, including the ones generated by hydra runs. We ignore the yaml files present in any `.hydra` directory.
```
Comment (omry) [inline website/docs/development/contributing.md:44]: tmi.

```suggestion
 - `yamllint .` : Ensures that yaml files are well formatted and consistent
```

Comment (omry) [reply website/docs/development/contributing.md:44]: Also, you said:
> When running locally, `yamllint` will run on all the yaml files, including the ones generated by hydra runs. We ignore the yaml files present in any `.hydra` directory.

This is not clear to me, besides .hydra, what other yaml files are generated?

Comment (shagunsodhani) [reply website/docs/development/contributing.md:44]: I think nothing else.

Comment (shagunsodhani) [reply website/docs/development/contributing.md:44]: I am removing the text you highlighted. Should I make a towncrier entry as well ?

File: .yamllint [hunk 5]
```diff
@@ -0,0 +1,36 @@
+---
+
```
[the remark this answers is not in the document]
Comment (shagunsodhani) [reply .yamllint:2]: Thanks for catching this!

File: website/docs/development/contributing.md (hunk 4 above)
Comment (omry) [reply website/docs/development/contributing.md:44]: still seeing the text.

**Pull request merged**